[llvm] Fix the MCSubtargetInfo used for module-level assembly.

chrisnc opened the pull request:
Provide both the default target CPU and default target features from the
module's context, rather than empty strings.

Fixes #61991.

chrisnc committed:
[llvm] Fix the MCSubtargetInfo used for module-level assembly.

Provide both the default target CPU and default target features from the
module's context, rather than empty strings.

Fixes #61991.

chrisnc commented:
`test.c`

```c
asm("vmsr fpexc, r0");
```

Without fix:

```bash
$ bin/clang -target arm-none-eabi test.c -c -flto=thin -mcpu=cortex-r4
<inline asm>:1:1: error: instruction requires: VFP2
    1 | vmsr fpexc, r0
      | ^
1 error generated.

$ bin/clang -target arm-none-eabi test.c -c -flto=thin -mcpu=cortex-r5
<inline asm>:1:1: error: instruction requires: VFP2
    1 | vmsr fpexc, r0
      | ^
1 error generated.
```

With fix, using `-mcpu=cortex-r5` (which implies FPU support, while `cortex-r4` does not), does not emit an error:
```bash
$ bin/clang -target arm-none-eabi test.c -c -flto=thin -mcpu=cortex-r4
<inline asm>:1:1: error: instruction requires: VFP2
    1 | vmsr fpexc, r0
      | ^
1 error generated.

$ bin/clang -target arm-none-eabi test.c -c -flto=thin -mcpu=cortex-r5
```

chrisnc commented:
Credit to https://github.com/llvm/llvm-project/issues/61991#issuecomment-2207148632 @Dirbaio for finding where the problem was.

Hopefully this is not a breaking change for anything and the fix is this simple...

efriedma-quic commented:
The idea here makes sense, but the API in question is likely to change, so I'd rather not add new uses of that API until we've sorted that out.  See https://discourse.llvm.org/t/functions-generated-by-function-createwithdefaultattr-should-respect-target-features/79838 .

MaskRay reviewed (changes requested):
Request changes since the ABI will change. My summary:

<https://github.com/llvm/llvm-project/pull/96721> implemented a strategy by storing the target features in `LLVMContext`, which is not serialized.
`Function::createWithDefaultAttr` will load the target features from `LLVMContext`.

While 96721 works well, there is a drawback that the generate module from the LTO pre-link compilation does not contain sufficient information.
Running `opt` without the correct target features would not mimick Clang .
To address this drawback, we can use module flags metadata that is dropped during merging.

chrisnc commented:
Oh! I did not notice that the interfaces I'm using only appeared a week ago. Thanks for the additional context.

> use module flags metadata

Is this something I can do now at this call-site, or something that needs more iteration on what the correct way to store/retrieve this information from various phases should be?

chrisnc commented:
Hi @MaskRay, is https://github.com/llvm/llvm-project/pull/98673 the confirmed direction yet? If so, I'll watch for that to merge and then re-visit this.

MaskRay commented:
> Hi @MaskRay, is #98673 the confirmed direction for how this info will be handled? If so, I'll watch for that to merge and then re-visit this.

I think so. Thanks  for waiting :)

chrisnc commented:
Hi @MaskRay, I've updated this branch to resolve conflicts and added a test. Can I ask that this be considered now, even though https://github.com/llvm/llvm-project/pull/98673 is still open? The change is very small and solves a real problem that people keep encountering, and should be easy to refactor into the right shape when the intended direction for target feature tracking is ready to land.

nikic reviewed (changes requested):
Isn't https://github.com/llvm/llvm-project/pull/174995 the agreed upon direction to fix this?

chrisnc commented:
Bump. @nikic

chrisnc commented:
Bump. @nikic @MaskRay 
Is there anything still blocking this?

chrisnc commented:
Bump.

MaskRay commented:
> Bump.

Requires @nikic's approval
